// @wix/form-fields/dist/esm/ui/text-area-field/text-area-field-wut.js
import classNames from 'classnames';
import React, {
    useCallback
} from 'react';
import {
    TextArea as TPATextArea
} from 'wix-ui-tpa/cssVars';
import {
    FIELD_TYPES
} from '../../constants';
import {
    ForcedState,
    useForcedState
} from '../../contexts/forced-state-context';
import {
    FIELD_ERROR_CLASS,
    useError
} from '../../field-error';
import {
    useFieldAttributes
} from '../../services/field-attributes';
import {
    FieldDescriptionWUT
} from '../field-description';
import {
    FieldErrorWUT
} from '../field-error';
import {
    FieldLabelWUT
} from '../field-label';
import {
    useTpaStyleParams
} from '../use-tpa-style-params';
import {
    classes
} from './text-area-field-wut.st.css';
export const TextAreaFieldWUT = _ref => {
    let {
        id,
        target,
        label,
        hideLabel,
        placeholder,
        required,
        value,
        onChange,
        onFocus,
        onBlur,
        error,
        customErrors,
        disabled = false,
        description,
        fieldType = FIELD_TYPES.TEXT_AREA,
        FieldLayout
    } = _ref;
    const {
        labelId,
        inputId,
        errorId,
        descriptionId,
        ariaDescribedBy
    } = useFieldAttributes(id);
    const {
        hasError,
        errorMessage
    } = useError(error, customErrors, fieldType);
    const forcedHover = useForcedState().includes(ForcedState.INPUT_FIELDS_HOVER);
    const {
        inputTheme,
        isBorderStyleNone
    } = useTpaStyleParams();
    const handleChange = useCallback(event => {
        const newValue = event.target.value;
        onChange(newValue);
    }, [onChange]);
    return /*#__PURE__*/ React.createElement(FieldLayout, {
        fieldId: id,
        renderLabel: () => /*#__PURE__*/ React.createElement(FieldLabelWUT, {
            labelId: labelId,
            htmlFor: inputId,
            required: required,
            label: label,
            hideLabel: hideLabel
        }),
        renderInput: () => /*#__PURE__*/ React.createElement(TPATextArea, {
            id: inputId,
            error: hasError,
            newErrorMessage: true,
            className: classNames(classes.textArea, {
                [FIELD_ERROR_CLASS]: hasError,
                [classes.noBorder]: isBorderStyleNone,
                [classes.hover]: forcedHover
            }),
            ariaLabel: label ? ? '',
            value: value ? String(value) : '',
            onChange: handleChange,
            onFocus: onFocus,
            onBlur: onBlur,
            disabled: disabled,
            placeholder: placeholder,
            theme: inputTheme,
            "aria-describedby": ariaDescribedBy
        }),
        renderDescription: () => /*#__PURE__*/ React.createElement(React.Fragment, null, /*#__PURE__*/ React.createElement(FieldErrorWUT, {
            target: target,
            hasError: hasError,
            errorMessage: errorMessage,
            id: errorId
        }), /*#__PURE__*/ React.createElement(FieldDescriptionWUT, {
            description: description,
            id: descriptionId
        }))
    });
};
