// ricos-viewer-next/dist/es/getClassNames-2d778f5a.js
import {
    upperFirst,
    camelCase
} from 'lodash';

var styles = {
    "alignLeft": "m5OWB",
    "textWrapNowrap": "jlTZm",
    "alignRight": "lZTI-",
    "alignCenter": "lKCNC",
    "sizeFullWidth": "p4aW6",
    "renderedInTable": "NCE2l",
    "sizeOriginal": "gI7my",
    "sizeContent": "RLSxb",
    "sizeSmall": "KWORd",
    "sizeSmallCenter": "mmqsA",
    "flex": "JRyxQ",
    "atomic": "yirMQ",
    "anchor": "R3W7R",
    "embed": "IetyN",
    "horizontalScrollbar": "lmUPr",
    "pluginContainerMobile": "_5VqLH"
};

const getContainerData = (node) => {
    var _a;
    const data = `${camelCase(node.type)}Data`;
    return (_a = node == null ? void 0 : node[data]) == null ? void 0 : _a.containerData;
};
const getAlignmentClassName = (containerData = {}) => {
    const {
        alignment
    } = containerData;
    if (!alignment) {
        return "";
    }
    const key = `align${upperFirst(alignment.toLowerCase())}`;
    return styles[key];
};
const getSizeClassName = (containerData) => {
    const {
        width
    } = containerData;
    const {
        size
    } = width || {};
    if (!size) {
        return "";
    }
    const key = `size${upperFirst(camelCase(size.toLowerCase()))}`;
    return styles[key];
};
const getTextWrapClassName = (containerData = {}) => {
    const {
        textWrap
    } = containerData;
    if (textWrap) {
        return "";
    }
    return styles.textWrapNowrap;
};
const alignmentClassName = (node) => {
    return getAlignmentClassName(getContainerData(node));
};
const sizeClassName = (node) => {
    return getSizeClassName(getContainerData(node) || {});
};
const textWrapClassName = (node) => {
    return getTextWrapClassName(getContainerData(node));
};

export {
    getSizeClassName as a, getTextWrapClassName as b, alignmentClassName as c, sizeClassName as d, getAlignmentClassName as g, styles as s, textWrapClassName as t
};
