// @wix/form-fields/dist/esm/ui/form/components/form/nested-form-field.js
import _extends from "@babel/runtime/helpers/extends";
import {
    groupErrorsByTarget
} from '@wix/form-validator';
import React, {
    useCallback,
    useMemo
} from 'react';
import {
    useExternalData
} from '../../contexts/external-data-context';
import {
    useFormFields
} from '../../contexts/form-fields-context';
export const NestedFormField = _ref => {
    let {
        field,
        errors,
        values,
        onChange,
        FieldLayout
    } = _ref;
    const componentsById = useFormFields();
    const externalData = useExternalData();
    const fieldValue = values[field.target];
    const errorsByTarget = useMemo(() => groupErrorsByTarget(errors), [errors]);
    const fieldErrors = errorsByTarget == null ? void 0 : errorsByTarget[field.target];
    const fieldError = fieldErrors == null ? void 0 : fieldErrors[0];
    const handleChange = useCallback(value => {
        if (!value) {
            value = null;
        }
        const newValues = {
            ...values,
            [field.target]: value
        };
        onChange == null || onChange(newValues);
    }, [onChange, values, field]);
    const FieldComponent = componentsById[field.fieldType];
    if (!FieldComponent) {
        return null;
    }
    return /*#__PURE__*/ React.createElement(FieldComponent, _extends({}, field.properties, {
        onChange: handleChange,
        error: fieldError,
        errors: fieldErrors,
        value: fieldValue,
        id: field.id,
        target: field.target,
        path: field.path,
        externalData: externalData,
        fieldType: field.fieldType,
        FieldLayout: FieldLayout
    }));
};
