// @wix/form-fields/dist/esm/ui/ricos-viewer/ricos-viewer.js
import React, {
    useMemo
} from 'react';
import {
    pluginLink
} from 'wix-rich-content-plugin-link/viewer';
import {
    pluginLineSpacing
} from 'wix-rich-content-plugin-line-spacing/viewer';
import styles from './ricos-viewer.scss';
import classNames from 'classnames';
import {
    useStyleParams
} from '../../contexts/style-params-context';
import {
    getStyleParamsTheme
} from './theme-utils';
import {
    useConfig
} from '../../contexts/config-context';
import {
    useWixRicosViewer
} from '../../contexts/ricos-viewer-context';
import {
    useBi
} from '../../contexts/bi-context';
const RicosViewer = _ref => {
    let {
        content,
        className,
        customTheme
    } = _ref;
    const WixRicosViewer = useWixRicosViewer();
    const styleParams = useStyleParams();
    const {
        locale
    } = useConfig();
    const bi = useBi();
    const ricosTheme = useMemo(() => getStyleParamsTheme({
        styleParams,
        customTheme
    }), [customTheme, styleParams]);
    if (!content) {
        return null;
    }
    return /*#__PURE__*/ React.createElement("div", {
        className: classNames(styles.container, className)
    }, /*#__PURE__*/ React.createElement(WixRicosViewer, {
        biSettings: {
            consumer: 'Forms',
            platform: 'Livesite',
            postId: '',
            containerId: '',
            logger: bi
        },
        instance: "",
        theme: ricosTheme,
        content: content,
        plugins: [pluginLink(), pluginLineSpacing()],
        locale: locale
    }));
};
export default RicosViewer;
