// @wix/form-fields/dist/esm/ui/ricos-viewer/theme-utils.js
const getIsStyleParam = styleParams => Object.keys(styleParams.booleans ? ? {}).length > 0 || Object.keys(styleParams.numbers ? ? {}).length > 0 || Object.keys(styleParams.colors ? ? {}).length > 0 || Object.keys(styleParams.fonts ? ? {}).length > 0;
export const getStyleParamsTheme = _ref => {
    let {
        customTheme,
        styleParams
    } = _ref;
    const styleParamDefined = getIsStyleParam(styleParams);
    if (styleParamDefined) {
        return {
            palette: {
                type: 'rgb',
                bgColor: 'var(--wix-forms-formInputBackgroundColor)',
                textColor: 'var(--wix-forms-formParagraphColor)',
                actionColor: 'var(--wix-forms-formLinkColor)'
            },
            ...customTheme
        };
    }
    return defaultTheme;
};
const defaultTheme = {
    customStyles: {
        h1: {
            fontSize: '28px'
        },
        h2: {
            fontSize: '20px'
        },
        p: {
            fontSize: '16px'
        }
    }
};
export const getHeaderAndParagraphCustomStyles = () => {
    return {
        h1: {
            fontSize: 'var(--wix-forms-formHeaderOneFont-size)',
            fontFamily: 'var(--wix-forms-formHeaderOneFont-family)',
            color: 'rgb(var(--wix-forms-formHeaderOneColor))'
        },
        h2: {
            fontSize: 'var(--wix-forms-formHeaderTwoFont-size)',
            fontFamily: 'var(--wix-forms-formHeaderTwoFont-family)',
            color: 'rgb(var(--wix-forms-formHeaderTwoColor))'
        },
        p: {
            fontSize: 'var(--wix-forms-formParagraphFont-size)',
            fontFamily: 'var(--wix-forms-formParagraphFont-family)',
            color: 'rgb(var(---wix-forms-formParagraphColor))'
        }
    };
};
