// @wix/form-fields/dist/esm/ui/number-input-field/number-input-field-wut.js
import classNames from 'classnames';
import React, {
    useCallback
} from 'react';
import {
    TextField as TPATextField
} from 'wix-ui-tpa/cssVars';
import {
    FIELD_TYPES
} from '../../constants';
import {
    ForcedState,
    useForcedState
} from '../../contexts/forced-state-context';
import {
    FIELD_ERROR_CLASS,
    useError
} from '../../field-error';
import {
    getNumberInputError,
    isNumberInputValid
} from '../../fields/number-input/number-input/number-input-utils';
import {
    useFocus
} from '../../hooks/use-focus';
import {
    useFieldAttributes
} from '../../services/field-attributes';
import {
    FieldDescriptionWUT
} from '../field-description';
import {
    FieldErrorWUT
} from '../field-error';
import {
    FieldLabelWUT
} from '../field-label';
import {
    classes
} from './number-input-field-wut.st.css';
import {
    useTpaStyleParams
} from '../use-tpa-style-params';
export const NumberInputFieldWUT = _ref => {
    var _validation$number, _validation$number2, _validation$number3;
    let {
        id,
        target,
        value,
        validation,
        required,
        label,
        hideLabel,
        disabled,
        placeholder,
        onChange,
        onBlur,
        description,
        onFocus,
        error,
        customErrors,
        fieldType = FIELD_TYPES.NUMBER_INPUT,
        additionalInfo,
        FieldLayout
    } = _ref;
    const {
        labelId,
        inputId,
        errorId,
        descriptionId,
        ariaDescribedBy
    } = useFieldAttributes(id);
    const _error = getNumberInputError(validation, error);
    const {
        hasError,
        errorMessage
    } = useError(_error, customErrors, fieldType);
    const multipleOf = (validation == null || (_validation$number = validation.number) == null ? void 0 : _validation$number.multipleOf) || 1;
    const maximum = validation == null || (_validation$number2 = validation.number) == null ? void 0 : _validation$number2.maximum;
    const minimum = validation == null || (_validation$number3 = validation.number) == null ? void 0 : _validation$number3.minimum;
    const forcedHover = useForcedState().includes(ForcedState.INPUT_FIELDS_HOVER);
    const {
        inputTheme,
        isBorderStyleNone
    } = useTpaStyleParams();
    const numberOnChange = useCallback(event => {
        const rawValue = event.target.value;
        if (rawValue === '') {
            // allow clear input
            onChange(rawValue);
            return;
        }
        const newValue = Number(rawValue);
        const validValue = isNumberInputValid(newValue, multipleOf) ? newValue : value;
        onChange(validValue);
    }, [onChange, multipleOf, value]);
    const {
        handleFocus,
        handleBlur,
        containerRef
    } = useFocus({
        onBlur,
        onFocus
    });
    return /*#__PURE__*/ React.createElement(FieldLayout, {
        fieldId: id,
        renderLabel: () => /*#__PURE__*/ React.createElement(FieldLabelWUT, {
            labelId: labelId,
            htmlFor: inputId,
            required: required,
            label: label,
            hideLabel: hideLabel
        }),
        renderInput: () => /*#__PURE__*/ React.createElement("div", {
            ref: containerRef,
            onFocus: handleFocus,
            onBlur: handleBlur
        }, /*#__PURE__*/ React.createElement(TPATextField, {
            id: inputId,
            type: "number",
            error: hasError,
            newErrorMessage: true,
            value: typeof value === 'number' ? String(value) : value,
            className: classNames(classes.textField, {
                [FIELD_ERROR_CLASS]: hasError,
                [classes.noBorder]: isBorderStyleNone,
                [classes.hover]: forcedHover
            }),
            onChange: numberOnChange,
            disabled: disabled,
            placeholder: placeholder,
            min: minimum ? ? undefined,
            max: maximum ? ? undefined,
            "aria-label": label,
            step: multipleOf,
            theme: inputTheme,
            "aria-describedby": ariaDescribedBy
        })),
        renderDescription: () => /*#__PURE__*/ React.createElement(React.Fragment, null, /*#__PURE__*/ React.createElement(FieldErrorWUT, {
            target: target,
            hasError: hasError,
            errorMessage: errorMessage,
            id: errorId
        }), /*#__PURE__*/ React.createElement(FieldDescriptionWUT, {
            description: description,
            id: descriptionId
        }), additionalInfo)
    });
};
