// @wix/form-viewer/dist/esm/tpa/controller/error-monitor.js
import {
    getAppParams
} from './app-params';
import {
    getURL,
    getIsSEO,
    getIsSSR
} from './wix-code-api';
export const getArtifactVersion = () => {
    if (process.env.NODE_ENV === 'production') {
        return process.env.ARTIFACT_VERSION;
    }
    return 'local-development';
};
export const createErrorMonitor = ({
    essentials,
    wixCodeApi,
    appParams,
}) => {
    const {
        appName,
        appDefinitionId,
        appVersion
    } = getAppParams(appParams);
    // HACK: essentials.createErrorMonitor is missing in Yoshi testkit.
    if (process.env.NODE_ENV === 'test') {
        const errorMonitor = {
            addBreadcrumb: () => {},
            captureException: () => 'captureException',
            captureMessage: () => 'captureMessage',
        };
        return errorMonitor;
    }
    const errorMonitor = essentials.createErrorMonitor({
        dsn: 'https://3beae28b544b43e8976029ab465f9ba9@sentry-next.wixpress.com/2049',
        environment: 'Controller',
        version: getArtifactVersion(),
        tags: {
            appName,
            appDefinitionId,
            appVersion,
            isSSR: getIsSEO(wixCodeApi),
            isSEO: getIsSSR(wixCodeApi),
            url: getURL(wixCodeApi),
        },
    });
    return errorMonitor;
};
