// @wix/ambassador-forms-assistant-v2-thread/build/es/http.impl.js
import {
    toURLSearchParams
} from '@wix/metro-runtime';
import {
    serializer
} from '@wix/metro-runtime/ambassador';
import {
    resolveUrl
} from '@wix/metro-runtime';
var _createThreadRequest = {
    thread: '_thread'
};
var _createThreadResponse = {
    thread: '_thread'
};
var _getThreadByFormIdRequest = {};
var _getThreadByFormIdResponse = {
    thread: '_thread'
};
var _getThreadRequest = {};
var _getThreadResponse = {
    thread: '_thread'
};
var _thread = {
    createdDate: 'google.protobuf.Timestamp',
    updatedDate: 'google.protobuf.Timestamp',
};

function resolveWixFormsAssistantV2ThreadServiceUrl(opts) {
    var domainToMappings = {
        _: [{
            srcPath: '/assistant-server',
            destPath: '',
        }, ],
        'forms.wixapps.net': [{
            srcPath: '/assistant-server',
            destPath: '',
        }, ],
        'apps._base_domain_': [{
            srcPath: '/assistant-server',
            destPath: '',
        }, ],
    };
    return resolveUrl(Object.assign(opts, {
        domainToMappings: domainToMappings
    }));
}
/** Creates a thread. */
export function createThread(payload) {
    var _a = serializer(_createThreadRequest, {
            _thread: _thread
        }),
        toReq = _a.toJSON,
        fromReq = _a.fromJSON;
    var fromRes = serializer(_createThreadResponse, {
        _thread: _thread
    }).fromJSON;

    function __createThread(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.forms.assistant.v2.thread',
            method: 'POST',
            methodFqn: 'wix.forms.assistant.v2.ThreadService.CreateThread',
            url: resolveWixFormsAssistantV2ThreadServiceUrl({
                protoPath: '/v2/threads',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __createThread.fromReq = fromReq;
    __createThread.__isAmbassador = true;
    return __createThread;
}
/** Retrieves a thread. */
export function getThread(payload) {
    var _a = serializer(_getThreadRequest, {}),
        toReq = _a.toJSON,
        fromReq = _a.fromJSON;
    var fromRes = serializer(_getThreadResponse, {
        _thread: _thread
    }).fromJSON;

    function __getThread(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.forms.assistant.v2.thread',
            method: 'GET',
            methodFqn: 'wix.forms.assistant.v2.ThreadService.GetThread',
            url: resolveWixFormsAssistantV2ThreadServiceUrl({
                protoPath: '/v2/threads/{threadId}',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __getThread.fromReq = fromReq;
    __getThread.__isAmbassador = true;
    return __getThread;
}
/** Retrieves the latest thread for the given form ID. */
export function getThreadByFormId(payload) {
    var _a = serializer(_getThreadByFormIdRequest, {}),
        toReq = _a.toJSON,
        fromReq = _a.fromJSON;
    var fromRes = serializer(_getThreadByFormIdResponse, {
        _thread: _thread,
    }).fromJSON;

    function __getThreadByFormId(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.forms.assistant.v2.thread',
            method: 'GET',
            methodFqn: 'wix.forms.assistant.v2.ThreadService.GetThreadByFormId',
            url: resolveWixFormsAssistantV2ThreadServiceUrl({
                protoPath: '/v2/threads/form-id/{formId}',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __getThreadByFormId.fromReq = fromReq;
    __getThreadByFormId.__isAmbassador = true;
    return __getThreadByFormId;
}
