// @wix/form-viewer/dist/esm/tpa/widget/form-viewer-lean.js
import React, {
    forwardRef
} from 'react';
import Widget from './widget';
import ErrorBoundary from './error-boundary';
import ControllerError from './controller-error';
import {
    useFormViewerContext
} from './form-viewer-context';
const FormViewerLean = forwardRef(({
    i18n,
    formId,
    stepId,
    values,
    onChange,
    onStepIdChange,
    errors,
    onValidate,
    onError,
    useSiteTheme,
    externalData,
    overrides,
    bi,
    biFormFillSessionId,
    forcedState,
    httpClient,
    style,
    onSubmit,
    onSubmitSuccess,
    onSubmitFailure,
    fields,
    WixRicosViewer,
}, forwardedRef) => {
    const {
        host,
        controller
    } = useFormViewerContext();
    return (React.createElement(ErrorBoundary, {
            host: host,
            wfConfig: controller.wfConfig,
            onError: onError
        },
        React.createElement(ControllerError, {
                wfConfig: controller.wfConfig,
                onError: onError
            },
            React.createElement(Widget, {
                httpClient: httpClient,
                i18n: i18n,
                bi: bi,
                biFormFillSessionId: biFormFillSessionId,
                formId: formId,
                values: values,
                stepId: stepId,
                onChange: onChange,
                onStepIdChange: onStepIdChange,
                errors: errors,
                onValidate: onValidate,
                forwardedRef: forwardedRef,
                controller: controller,
                onError: onError,
                useSiteTheme: useSiteTheme,
                externalData: externalData,
                overrides: overrides,
                forcedState: forcedState,
                style: style,
                fields: fields,
                WixRicosViewer: WixRicosViewer,
                onSubmit: onSubmit,
                onSubmitSuccess: onSubmitSuccess,
                onSubmitFailure: onSubmitFailure
            }))));
});
export default FormViewerLean;
