// @wix/ambassador-forms-assistant-v2-message/build/es/http.impl.js
import {
    serializer
} from '@wix/metro-runtime/ambassador';
import {
    resolveUrl
} from '@wix/metro-runtime';
var _choices = {
    numeric: '_numericChoices'
};
var _createMessageRequest = {
    message: '_message'
};
var _createMessageResponse = {};
var _message = {
    createdDate: 'google.protobuf.Timestamp',
    updatedDate: 'google.protobuf.Timestamp',
    context: '_messageContext',
};
var _messageContext = {
    fieldChoices: '_choices'
};
var _numericChoices = {
    value: 'google.protobuf.DoubleValue'
};
var _queryMessagesRequest = {};
var _queryMessagesResponse = {
    messages: '_message'
};

function resolveWixFormsAssistantV2MessageServiceUrl(opts) {
    var domainToMappings = {
        _: [{
            srcPath: '/assistant-server',
            destPath: '',
        }, ],
        'forms.wixapps.net': [{
            srcPath: '/assistant-server',
            destPath: '',
        }, ],
        'apps._base_domain_': [{
            srcPath: '/assistant-server',
            destPath: '',
        }, ],
    };
    return resolveUrl(Object.assign(opts, {
        domainToMappings: domainToMappings
    }));
}
/** Creates a message in a thread or specifies the form data that is already filled in. */
export function createMessage(payload) {
    var _a = serializer(_createMessageRequest, {
            _choices: _choices,
            _message: _message,
            _messageContext: _messageContext,
            _numericChoices: _numericChoices
        }),
        toReq = _a.toJSON,
        fromReq = _a.fromJSON;
    var fromRes = serializer(_createMessageResponse, {}).fromJSON;

    function __createMessage(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.forms.assistant.v2.message',
            method: 'POST',
            methodFqn: 'wix.forms.assistant.v2.MessageService.CreateMessage',
            url: resolveWixFormsAssistantV2MessageServiceUrl({
                protoPath: '/v2/threads/{threadId}/messages',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __createMessage.fromReq = fromReq;
    __createMessage.__isAmbassador = true;
    return __createMessage;
}
/** Query messages in a thread. */
export function queryMessages(payload) {
    var _a = serializer(_queryMessagesRequest, {}),
        toReq = _a.toJSON,
        fromReq = _a.fromJSON;
    var fromRes = serializer(_queryMessagesResponse, {
        _choices: _choices,
        _message: _message,
        _messageContext: _messageContext,
        _numericChoices: _numericChoices,
    }).fromJSON;

    function __queryMessages(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.forms.assistant.v2.message',
            method: 'POST',
            methodFqn: 'wix.forms.assistant.v2.MessageService.QueryMessages',
            url: resolveWixFormsAssistantV2MessageServiceUrl({
                protoPath: '/v2/threads/{threadId}/messages/query',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __queryMessages.fromReq = fromReq;
    __queryMessages.__isAmbassador = true;
    return __queryMessages;
}
