// @wix/ambassador-forms-v4-restrictions/build/es/http.impl.js
import {
    toURLSearchParams
} from '@wix/metro-runtime';
import {
    serializer
} from '@wix/metro-runtime/ambassador';
import {
    resolveUrl
} from '@wix/metro-runtime';
var _getRestrictionsRequest = {};
var _getRestrictionsResponse = {};

function resolveWixFormsV4RestrictionsServiceUrl(opts) {
    var domainToMappings = {
        'bo._base_domain_': [{
            srcPath: '/form-app-service',
            destPath: '',
        }, ],
        'www.wixapis.com': [{
                srcPath: '/_api/form-app-service',
                destPath: '',
            },
            {
                srcPath: '/form-app-service',
                destPath: '',
            },
        ],
        'manage._base_domain_': [{
            srcPath: '/form-app-service',
            destPath: '',
        }, ],
        'api._api_base_domain_': [{
            srcPath: '/form-app-service',
            destPath: '',
        }, ],
        _: [{
            srcPath: '/_api/form-app-service',
            destPath: '',
        }, ],
        'editor.wixapps.net': [{
            srcPath: '/_api/form-app-service',
            destPath: '',
        }, ],
        'editor._base_domain_': [{
            srcPath: '/_api/form-app-service',
            destPath: '',
        }, ],
        'blocks._base_domain_': [{
            srcPath: '/_api/form-app-service',
            destPath: '',
        }, ],
        'create.editorx': [{
            srcPath: '/_api/form-app-service',
            destPath: '',
        }, ],
        'apps._base_domain_': [{
            srcPath: '/form-app-service',
            destPath: '',
        }, ],
        'www._base_domain_': [{
            srcPath: '/_api/form-app-service',
            destPath: '',
        }, ],
        '*.dev.wix-code.com': [{
            srcPath: '/_api/form-app-service',
            destPath: '',
        }, ],
    };
    return resolveUrl(Object.assign(opts, {
        domainToMappings: domainToMappings
    }));
}
/** Get premium restrictions. */
export function getRestrictions(payload) {
    var _a = serializer(_getRestrictionsRequest, {}),
        toReq = _a.toJSON,
        fromReq = _a.fromJSON;
    var fromRes = serializer(_getRestrictionsResponse, {}).fromJSON;

    function __getRestrictions(_a) {
        var host = _a.host;
        var serializedData = toReq(payload);
        var metadata = {
            entityFqdn: 'wix.forms.v4.restrictions',
            method: 'GET',
            methodFqn: 'wix.forms.v4.RestrictionsService.GetRestrictions',
            url: resolveWixFormsV4RestrictionsServiceUrl({
                protoPath: '/v4/restrictions',
                data: serializedData,
                host: host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __getRestrictions.fromReq = fromReq;
    return __getRestrictions;
}
